onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -227,22 +227,7 @@ CUDAExecutionProvider::CUDAExecutionProvider(const CUDAExecutionProviderInfo& in
 }
 
 CUDAExecutionProvider::~CUDAExecutionProvider() {
-  auto cpu_alloc = GetAllocator(DEFAULT_CPU_ALLOCATOR_DEVICE_ID, OrtMemTypeCPU);
-  {
-    std::lock_guard<OrtMutex> lock(deferred_release_cpu_ptr_mutex_);
-    auto it = deferred_release_cpu_ptr_.begin();
-    while (it != deferred_release_cpu_ptr_.end()) {
-      auto& e = it->first;
-      auto& v = it->second;
-      if (v.recorded)
-        CUDA_CALL_THROW(cudaEventSynchronize(e));
-      for (auto p : v.cpu_ptrs) {
-        cpu_alloc->Free(p);
-      }
-      CUDA_CALL_THROW(cudaEventDestroy(e));
-      it = deferred_release_cpu_ptr_.erase(it);
-    }
-  }
+  EnqueueDeferredRelease();
```

Comment:
do we have lifetime issue for the allocator ptr? in the cuda call back you enqueued, it rely on the allocator ptr to release the buffer. but here after Cuda EP get deconstructed, the allocator may be gone if the cuda callback doesn't hold the reference count for the allocator ptr.

Reply:
Yea.. Let's also cache allocator (type: `using AllocatorPtr = std::shared_ptr<IAllocator>;`). It should be fine since it's a pointer.

Reply:
i see...